File: tests/java_helics/TestMessageFederate.java
import java.lang.reflect.Method;
import java.util.*;

import com.java.helics.SWIGTYPE_p_void;
import com.java.helics.helics;
import com.java.helics.federate_state;
import com.java.helics.message_t;
import com.java.helics.helics_filter_type_t;
import com.java.helics.helics_time_properties;

public class TestMessageFederate {

	public static SWIGTYPE_p_void createMessageFederate(final String fedName, final double timeDelta) {
		String fedinitstring = "--broker=mainbroker --federates=1";
		// Create Federate Info object that describes the federate properties
		SWIGTYPE_p_void fi = helics.helicsCreateFederateInfo();
		// TIME_DELTA_PROPERTY = 137
		helics.helicsFederateInfoSetTimeProperty(fi, helics_time_properties.helics_time_property_time_delta.swigValue(), timeDelta);
		helics.helicsFederateInfoSetCoreTypeFromString(fi, "zmq");
		// Federate init string
		helics.helicsFederateInfoSetCoreInitString(fi, fedinitstring);
		SWIGTYPE_p_void msgFed = helics.helicsCreateMessageFederate(fedName, fi);
		// Wait for sometime

		return msgFed;
	}

	public static void test_message_federate_initialize(SWIGTYPE_p_void msgFederate) {
		int[] state = new int[1];
		federate_state status = helics.helicsFederateGetState(msgFederate);
		assert status.swigValue() == 0;

		// helics.helicsFederateEnterExecutingMode(msgFederate);

		status = helics.helicsFederateGetState(msgFederate);
		assert status.swigValue() == 2;
	}

	public static void test_message_federate_endpoint_registration(SWIGTYPE_p_void msgFederate) {
		federate_state status = helics.helicsFederateGetState(msgFederate);
		
		assert status.swigValue() == 0;
		SWIGTYPE_p_void epid1 = helics.helicsFederateRegisterEndpoint(msgFederate, "ep1", "");
		SWIGTYPE_p_void epid2 = helics.helicsFederateRegisterGlobalEndpoint(msgFederate, "ep2", "random");

		helics.helicsFederateEnterExecutingMode(msgFederate);
		assert status.swigValue() == 3;
		
		int maxLen = 100;
		// byte[] endpoint_name = new byte[100];
		String endpoint_name = helics.helicsEndpointGetName(epid1);

		// String ep = new String(endpoint_name);
		assert endpoint_name == "TestA Federate/ep1";

		endpoint_name = helics.helicsEndpointGetName(epid2);
		// String ep2 = new String(endpoint_name);
		assert endpoint_name == "ep2";

		String endpoint_type = helics.helicsEndpointGetType(epid1);
		// String et = new String(endpoint_type);
		assert endpoint_type == "";

		endpoint_type = helics.helicsEndpointGetType(epid2);
		// String et2 = new String(endpoint_type);
		assert endpoint_type == "random";

	}

	
	public static void main(String[] args) {
		String initString = "1 --name=mainbroker";
		// Create broker
		SWIGTYPE_p_void broker = helics.helicsCreateBroker("zmq", "", initString);
		int isConnected = helics.helicsBrokerIsConnected(broker);
		// if isConnected == 1:
		// Test message federate creation
		SWIGTYPE_p_void myFederate = TestMessageFederate.createMessageFederate("TestA Federate", 1);


		// Test endpoint registration
		test_message_federate_endpoint_registration(myFederate);

		//
		// Wait for sometime
		helics.helicsFederateFinalize(myFederate);
		federate_state state = helics.helicsFederateGetState(myFederate);
		assert state.swigValue() == 3;
		while (helics.helicsBrokerIsConnected(broker) == 1) {
			try {
				Thread.sleep(1000); // 1000 milliseconds is one second.
			} catch (InterruptedException ex) {
				Thread.currentThread().interrupt();
			}
		}
		helics.helicsFederateFree(myFederate);
		helics.helicsCloseLibrary();

		// Test
	}

}

File: tests/python/test_message_federate.py
import time
import pytest as pt
import helics as h


@pt.fixture
def mFed():
    initstring = "-f 1 --name=mainbroker"
    fedinitstring = "--broker=mainbroker --federates=1"
    deltat = 0.01

    h.helicsGetVersion()

    # Create broker #
    broker = h.helicsCreateBroker("zmq", "", initstring)

    isconnected = h.helicsBrokerIsConnected(broker)

    if isconnected == 1:
        pass

    # Create Federate Info object that describes the federate properties #
    fedinfo = h.helicsCreateFederateInfo()

    # Set Federate name #
    h.helicsFederateInfoSetCoreName(fedinfo, "CoreA Federate")

    # Set core type from string #
    h.helicsFederateInfoSetCoreTypeFromString(fedinfo, "zmq")

    # Federate init string #
    h.helicsFederateInfoSetCoreInitString(fedinfo, fedinitstring)

    # Set the message interval (timedelta) for federate. Note th#
    # HELICS minimum message time interval is 1 ns and by default
    # it uses a time delta of 1 second. What is provided to the
    # setTimedelta routine is a multiplier for the default timedelta.

    # Set one second message interval #
    h.helicsFederateInfoSetTimeProperty(fedinfo, h.helics_property_time_delta, deltat)

    h.helicsFederateInfoSetIntegerProperty(fedinfo, h.helics_property_int_log_level, 1)

    mFed = h.helicsCreateMessageFederate("TestA Federate", fedinfo)

    yield mFed

    h.helicsFederateFinalize(mFed)
    state = h.helicsFederateGetState(mFed)
    assert state == 3
    while h.helicsBrokerIsConnected(broker):
        time.sleep(1)

    h.helicsFederateInfoFree(fedinfo)
    h.helicsFederateFree(mFed)
    h.helicsCloseLibrary()


def test_message_federate_initialize(mFed):
    state = h.helicsFederateGetState(mFed)
    assert state == 0
    h.helicsFederateEnterExecutingMode(mFed)

    state = h.helicsFederateGetState(mFed)
    assert state == 2


def test_message_federate_endpoint_registration(mFed):
    epid1 = h.helicsFederateRegisterEndpoint(mFed, "ep1", None)
    epid2 = h.helicsFederateRegisterGlobalEndpoint(mFed, "ep2", "random")

    h.helicsFederateEnterExecutingMode(mFed)

    endpoint_name = h.helicsEndpointGetName(epid1)
    assert endpoint_name == "TestA Federate/ep1"

    endpoint_name = h.helicsEndpointGetName(epid2)
    assert endpoint_name == "ep2"

    endpoint_name = h.helicsEndpointGetType(epid1)
    assert endpoint_name == ""

    endpoint_name = h.helicsEndpointGetType(epid2)
    assert endpoint_name == "random"


def test_message_federate_send(mFed):
    epid1 = h.helicsFederateRegisterEndpoint(mFed, "ep1", None)
    epid2 = h.helicsFederateRegisterGlobalEndpoint(mFed, "ep2", "random")

    h.helicsFederateSetTimeProperty(mFed, h.helics_property_time_delta, 1.0)
    h.helicsFederateEnterExecutingMode(mFed)

    data = "random-data"

    h.helicsEndpointSendEventRaw(epid1, "ep2", data, 1.0)

    granted_time = h.helicsFederateRequestTime(mFed, 2.0)
    assert granted_time == 1.0

    res = h.helicsFederateHasMessage(mFed)
    assert res == 1

    res = h.helicsEndpointHasMessage(epid1)
    assert res == 0

    res = h.helicsEndpointHasMessage(epid2)
    assert res == 1

    message = h.helicsEndpointGetMessage(epid2)

    assert message.data == "random-data"
    assert message.length == 11
    assert message.original_dest == ""
    assert message.original_source == "TestA Federate/ep1"
    assert message.source == "TestA Federate/ep1"
    assert message.time == 1.0
